test/acvp_test.c

```diff
@@ -0,0 +1,1370 @@
+    EVP_CIPHER *cipher = NULL;
+    unsigned char out[256] = { 0 };
+
+    TEST_note("%s : %s", alg, enc ? "encrypt" : "decrypt");
+    if (!TEST_ptr(ctx = EVP_CIPHER_CTX_new())
+        || !TEST_ptr(cipher = EVP_CIPHER_fetch(libctx, alg, ""))
+        || !TEST_true(EVP_CipherInit_ex(ctx, cipher, NULL, key, iv, enc))
+        || !TEST_true(EVP_CIPHER_CTX_set_padding(ctx, 0))
+        || !TEST_true(EVP_CipherUpdate(ctx, out, &len, pt, pt_len))
+        || !TEST_true(EVP_CipherFinal_ex(ctx, out + len, &out_len)))
+        goto err;
+    out_len += len;
+    if (!TEST_mem_eq(out, out_len, ct, ct_len))
+        goto err;
+    ret = 1;
+err:
+    EVP_CIPHER_free(cipher);
+    EVP_CIPHER_CTX_free(ctx);
+    return ret;
+}
+
+static int cipher_enc_dec_test(int id)
+{
+    const struct cipher_st *tst = &cipher_enc_data[id];
+    const int enc = 1;
+
+    return TEST_true(cipher_enc(tst->alg, tst->pt, tst->pt_len,
+                                tst->key, tst->key_len,
+                                tst->iv, tst->iv_len,
+                                tst->ct, tst->ct_len, enc))
+           && TEST_true(cipher_enc(tst->alg, tst->ct, tst->ct_len,
+                                   tst->key, tst->key_len,
+                                   tst->iv, tst->iv_len,
+                                   tst->pt, tst->pt_len, !enc));
+}
+
+static int aes_ccm_enc_dec(const char *alg,
+                           const unsigned char *pt, size_t pt_len,
+                           const unsigned char *key, size_t key_len,
+                           const unsigned char *iv, size_t iv_len,
+                           const unsigned char *aad, size_t aad_len,
+                           const unsigned char *ct, size_t ct_len,
+                           const unsigned char *tag, size_t tag_len,
+                           int enc, int pass)
+{
+    int ret = 0;
+    EVP_CIPHER_CTX *ctx;
+    EVP_CIPHER *cipher = NULL;
+    int out_len, len;
+    unsigned char out[1024];
+
+    TEST_note("%s : %s : expected to %s", alg, enc ? "encrypt" : "decrypt",
+              pass ? "pass" : "fail");
+
+    if (!TEST_ptr(ctx = EVP_CIPHER_CTX_new())
+        || !TEST_ptr(cipher = EVP_CIPHER_fetch(libctx, alg, ""))
+        || !TEST_true(EVP_CipherInit_ex(ctx, cipher, NULL, NULL, NULL, enc))
+        || !TEST_true(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, iv_len,
+                                          NULL))
+        || !TEST_true(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, tag_len,
+                                          enc ? NULL : (void *)tag))
+        || !TEST_true(EVP_CipherInit_ex(ctx, NULL, NULL, key, iv, enc))
+        || !TEST_true(EVP_CIPHER_CTX_set_padding(ctx, 0))
+        || !TEST_true(EVP_CipherUpdate(ctx, NULL, &len, NULL, pt_len))
+        || !TEST_true(EVP_CipherUpdate(ctx, NULL, &len, aad, aad_len))
+        || !TEST_int_eq(EVP_CipherUpdate(ctx, out, &len, pt, pt_len), pass))
+        goto err;
+
+    if (!pass) {
```

Comment:
Maybe I'm misunderstanding the intent here. Shouldn't we always go ahead and verify the tag? But if pass is 0 we expect the tag verification to fail. Here we just don't do the tag verification at all.

Reply:
The test above is !TEST_int_eq(EVP_CipherUpdate(ctx, out, &len, pt, pt_len), pass))..
So doing a final and getting the tag is not the correct thing to do.

Reply:
Err...why are we expecting the update to fail?

Reply:
Ok so it happens since the last update passes the ciphertext and does the decrypt - because it does the decrypt is also does the tag check - Since the decrypted data should only be returned IF AND ONLY IF the tag is correct.

Reply:
Is this behaviour specific to CCM mode in general, or specific to CCM mode with the FIPS provider? I thought it was the case (generally) that we did not check the tag until the "Final" call. Otherwise how can we know that we've finished receiving data (there could be more "Update" calls to come with more ciphertext).

Reply:
I went back to 1_1_1 and debugged evp_test with CCM input. EVP_Final does nothing. There is even a comment saying final does nothing.
Note that for CCM the MAC is performed on the plaintext.
The test doesnt do fragging for CCM so it is basically a one shot operation.